lucille-core/src/test/java/com/kmwllc/lucille/stage/DictionaryLookupTest.java

```diff
@@ -80,7 +83,139 @@ public void testGetLegalProperties() throws StageException {
             "dest",
             "conditions",
             "class",
-            "dict_path"),
+            "dict_path",
+            "set_only",
+            "ignore_missing_source"),
         stage.getLegalProperties());
   }
+
+  @Test
+  public void testFileDoesNotExist() {
+    Throwable e = assertThrows(StageException.class, () -> factory.get("DictionaryLookupTest/missing_file.conf"));
+    // todo should consider propagating the underlying exception message to root (right now I see it in e.cause.target)
+    // assertEquals("File does not exist: classpath:DictionaryLookupTest/missing.txt", e.getMessage());
+    assertEquals("Failed to instantiate stage: com.kmwllc.lucille.stage.DictionaryLookup", e.getMessage());
+  }
```

Comment:
I would be very interested in investigating propagating error messages to the highest level for Lucille debugging. Can you help me draft a ticket for that?

Reply:
How are you looking for the exception stacktrace to be structured?

If you do `e.printStackTrace()` inside this test you currently get this:

`com.kmwllc.lucille.core.StageException: Failed to instantiate stage: com.kmwllc.lucille.stage.DictionaryLookup
	at com.kmwllc.lucille.stage.StageFactory.get(StageFactory.java:68)
	at com.kmwllc.lucille.stage.StageFactory.get(StageFactory.java:45)
	at com.kmwllc.lucille.stage.DictionaryLookupTest.lambda$testFileDoesNotExist$0(DictionaryLookupTest.java:94)
	at org.junit.Assert.assertThrows(Assert.java:1001)
	at org.junit.Assert.assertThrows(Assert.java:981)
	at com.kmwllc.lucille.stage.DictionaryLookupTest.testFileDoesNotExist(DictionaryLookupTest.java:94)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at com.kmwllc.lucille.stage.StageFactory.get(StageFactory.java:63)
	... 30 more
Caused by: com.kmwllc.lucille.core.StageException: File does not exist: classpath:DictionaryLookupTest/missing.txt
	at com.kmwllc.lucille.stage.DictionaryLookup.getFileReader(DictionaryLookup.java:210)
	at com.kmwllc.lucille.stage.DictionaryLookup.countLines(DictionaryLookup.java:223)
	at com.kmwllc.lucille.stage.DictionaryLookup.buildHashMap(DictionaryLookup.java:94)
	at com.kmwllc.lucille.stage.DictionaryLookup.<init>(DictionaryLookup.java:72)
	... 36 more
`

Reply:
formatting of comment got lost -- but the idea is that there's a top-level StageException that gets thrown when StageFactory fails to construct the new stage instance. Looking deeper into the causes we see that the stage constructor encountered a missing file. That's how it usually works -- top level exceptions have causes that you find further down

Reply:
i see the desired message like this `assertEquals("File does not exist: classpath:DictionaryLookupTest/missing.txt", e.getCause().getCause().getMessage())` 
It would be great if we say the cause message in the top-level exception.

Reply:
Thx for the comment on this. Let's continue this discussion elsewhere if we need to as it won't affect this particular PR.